### DISCUSSION TITLE
Max Parallel Ordered Classes or Tests

### DISCUSSION THREAD

**Discussion by fairking** [xunit/xunit#2667] [Question] [answered]
Hi guys, I have a problem with setting up the xunit tests execution.

Lets say I have 20 test classes. In all those classes I am using `IClassFixture<DbContext>` along with `TestCaseOrderer`.

By default if I execute the entire assembly, it starts gradually to run all test classes in parallel, as a result I get `Database ConnectionTimeoutException` because all test classes try to bootstrap `IClassFixture<DbContext>` all 20 at the same time (it creates a new database, migrates scheme and seeds some scenarios).

Is there any way to limit a maximum amount of classes to run in parallel (let say max 3 classes)?

Tried `MaxParallelThreads` but as far as I can see it does nothing. I know I can use `DisableTestParallelization` but it makes tests to run one by one and requires a lot of time to complete.

Anyone has some ideas?

BTW I am using xunit v2.4.2 and visual studio runner v2.4.5. Visual Studio 2022 v17.4.4.

**Comment by fairking** [comment, marked as answer]
My current temporary solution is as follows:
``` c#
public class DbFixture : IDisposable
{
    private static object _concurrentLock = new object();
    
    public DbContext Db { get; protected set; }
    
    public void Init(string databaseName)
    {
        lock (_concurrentLock)
        {
            Db = CreateMigrateAndSeedDatabase(databaseName) // Creates unique database for the current test case
        }
    }
}
```

But I still have timeouts sometimes due to running 20 test classes at the same time after the init takes place.

**Reply by bradwilson** [reply]
What you're looking for is a collection fixture, not a class fixture. That'll share a single instance of `DbFixture` among all the tests in the collection, and prohibit parallelism among them.

For more information, see [these docs](https://xunit.net/docs/running-tests-in-parallel#parallelism-in-test-frameworks) and [this sample](https://github.com/xunit/samples.xunit/tree/main/CollectionFixtureExample).

**Reply by fairking** [reply]
@bradwilson Thanks for your reply. I am not sure what exactly you mean by using collections for my tests, but the point is my tests must be executed with different scenarios and that is why I use `IClassFixture`.

For example:
``` c#
    [assembly: TestCaseOrderer("MyServices.Tests.AlphabeticalTestOrderer", "MyServices.Tests")]

    public class UserServiceTests : IClassFixture<DbFixture>
    {
        private readonly DbFixture _db;

        public UserServiceTests(DbFixture db)
        {
            _db = db;
        }

        [Fact]
        public void T00_InitScenario()
        {
            _db.Init("TestDb_UserServiceTests"); // Creates a new database with name "TestDb_UserServiceTests"

            // ... seed data ready for the current class test cases
        }

        [Fact]
        public void T01_CreateUser()
        {
            // Create user "John Smith"

            // Assert record is created
        }

        [Fact]
        public void T02_UpdateUser()
        {
            // Update user "John Smith"

            // Assert record is updated
        }

        [Fact]
        public void T03_DeleteUser()
        {
            // Delete user "John Smith"

            // Assert record is deleted
        }
    }

    public class InvoiceServiceTests : IClassFixture<DbFixture>
    {
        private readonly DbFixture _db;

        public InvoiceServiceTests(DbFixture db)
        {
            _db = db;
        }

        [Fact]
        public void T00_InitScenario()
        {
            _db.Init("TestDb_InvoiceServiceTests"); // Creates a new database with name "TestDb_InvoiceServiceTests"

            // ... seed data ready for the current class test cases
        }

        [Fact]
        public void T01_CreateInvoice()
        {
            // Create invoice XYZ

            // Assert record is created
        }

        [Fact]
        public void T02_AddInvoiceLines()
        {
            // Add invoice lines to the invoice XYZ

            // Assert record contain lines
        }

        [Fact]
        public void T03_RecalculateInvoiceBalance()
        {
            // Do some service operations and asserts
        }
    }
```
As you can see I am happy with this approach and I cannot see why do I need Collection Fixture
So the issue, again, is to limit the amount of test classes to run, because by default it executes all classes almost immediately and because of it bootstraps `_db.Init("TestDb_XXXX")` many scenarios at the same time, as a result it gets db connection timeout exception, because it tries to create 20 databases at the same time. I would like to limit this number of inits to lets say 3 or so in parallel.

So literally I would like to limit a number of classes run in parallel. Or let's say to limit a number of tests in parallel.

**Reply by bradwilson** [reply]
> I would like to limit a number of classes run in parallel.

We allow you to set a parallelization limit: https://xunit.net/docs/configuration-files#maxParallelThreads

I'm guessing that isn't going to actually do what you want, though. It limits the number of things that can be _run simultaneously_, but not the number of things that can be _started simultaneously_. None of your examples include `await`, but I'm guessing your real tests do, because sitting for an `await` to finish means we repurpose the thread to run something else.

We have no capability to limit the number of things that can be _started_ (and no plans for it, either).

**Reply by fairking** [reply]
> None of your examples include await, but I'm guessing your real tests do

Yes, all our db related tests are `async/await`

> We have no capability to limit the number of things that can be started

I cannot understand then let's say even if we limited (or not) a number of threads (not tests). If we have 1k of tests, does it mean they all will be started simultaneously at the same time? I just would like to confirm I understand it correctly. I assume `DisableTestParallelization = false`. If every test takes few minutes to run I could not imagine how would 1k be. It would be IO disk overhead because all those 1k tests will try to access the database/harddrive at the same time.

BTW if I switch my DbContext to [In-Memory OLTP](https://learn.microsoft.com/en-us/sql/relational-databases/in-memory-oltp/overview-and-usage-scenarios?view=sql-server-ver16) mode I got out of memory exception, while with `DisableTestParallelization` no errors so far.

The best option among with `MaxParallelThreads` would be `MaxParallelTests` taking into account ordered class fixtures as well. IMHO

**Reply by bradwilson** [reply]
> If we have 1k of tests, does it mean they all will be started simultaneously at the same time?

We will not necessarily start all 1000 tests at the same time. What we will do is start new tests when a thread becomes available; it may become available because a test has finished, or it might become available because a test is waiting via `await`.

As an extreme example, if you had 1000 tests that all just ran `await Task.Delay(5000);` then yes, it's very likely that all 1000 tests would get started, because as each test starts, it immediately yields the thread to sleep for 5 seconds. We could easily start 1000 tests within 5 seconds.

I hope that clarifies how it works.

**Reply by fairking** [reply]
Thanks @bradwilson for the clarification. I think the best workaround for me would be is to make all my tests as synchronous methods (no async/await anymore).

**Reply by bradwilson** [reply]
Yes, if my analysis of what's happening is correct, that would fix the issue for you.
